src/NUnitFramework/framework/Internal/Execution/WorkItemQueue.cs

```diff
@@ -117,22 +138,29 @@ public bool IsEmpty
         /// <param name="work">The WorkItem to process</param>
         public void Enqueue(WorkItem work)
         {
-            lock (_syncRoot)
+            do
```

Comment:
There is a large amount of published information on this issue, including MSDN.  Lockless thread safety is the current design goal for most contemporary programming.  It is higher performance, deadlock free, and much easier to debug.  The entire reason for the concurrent collections is to provide for lock free thread safe collections.  All of the regular collections can be make thread safe by using locks, but that can be fraught with problems and performance issues.  Using locks on enqueu and dequeu then requires that locks be used everywhere that any of the properties and methods of the class be used.  By using spinning and CAS ([https://en.wikipedia.org/wiki/Compare-and-swap]), there is no need for that.